skills/repo.go

```diff
@@ -157,6 +159,93 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
 	return nil
 }
 
+// AllSlots returns all skill slots of the given type.
+// If skillType is the empty string, all skill slots are returned.
+func (r *Repository) AllSlots(skillType string) []*Slot {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Slot
+	for _, slotsForSnap := range r.slots {
+		for _, slot := range slotsForSnap {
+			if skillType == "" || slot.Type == skillType {
+				result = append(result, slot)
+			}
+		}
+	}
+	sort.Sort(bySlotSnapAndName(result))
+	return result
+}
+
+// Slots returns the skill slots offered by the named snap.
+func (r *Repository) Slots(snapName string) []*Slot {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Slot
+	for _, slot := range r.slots[snapName] {
+		result = append(result, slot)
+	}
+	sort.Sort(bySlotSnapAndName(result))
+	return result
+}
+
+// Slot returns the specified skill slot from the named snap.
+func (r *Repository) Slot(snapName, slotName string) *Slot {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return r.slots[snapName][slotName]
+}
+
+// AddSlot adds a new slot to the repository.
+// Adding a slot with invalid name returns an error.
+// Adding a slot that has the same name and snap name as another slot returns an error.
+func (r *Repository) AddSlot(slot *Slot) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Reject snaps with invalid names
+	if err := snap.ValidateName(slot.Snap); err != nil {
+		return err
+	}
+	// Reject skill with invalid names
+	if err := ValidateName(slot.Name); err != nil {
+		return err
+	}
+	// TODO: ensure that apps are correct
+	t := r.types[slot.Type]
+	if t == nil {
+		return fmt.Errorf("cannot add slot, skill type %q is not known", slot.Type)
+	}
+	if _, ok := r.slots[slot.Snap][slot.Name]; ok {
+		return fmt.Errorf("cannot add slot, snap %q already has slot %q", slot.Snap, slot.Name)
```

Comment:
`"cannot add skill slot, ..."`

We should probably avoid speaking externally about slots without the context of skills.